diff -urp gcc-4.3.0-20080428-origplug/gcc/c-decl.c gcc-4.3.0-20080428/gcc/c-decl.c
--- gcc-4.3.0-20080428-origplug/gcc/c-decl.c	2008-07-10 14:19:27.000000000 +0200
+++ gcc-4.3.0-20080428/gcc/c-decl.c	2008-07-10 23:00:44.000000000 +0200
@@ -61,6 +61,7 @@ along with GCC; see the file COPYING3.  
 #include "except.h"
 #include "langhooks-def.h"
 #include "pointer-set.h"
+#include "tree-plugin-pass.h"
 
 /* In grokdeclarator, distinguish syntactic contexts of declarators.  */
 enum decl_context
@@ -5767,6 +5769,7 @@ finish_struct (tree t, tree fieldlist, t
 
   /* Finish debugging output for this type.  */
   rest_of_type_compilation (t, toplevel);
+  plugin_finish_struct (t);
 
   /* If we're inside a function proper, i.e. not file-scope and not still
      parsing parameters, then arrange for the size of a variable sized type
@@ -6792,6 +6795,7 @@ finish_function (void)
     {
       if (!decl_function_context (fndecl))
 	{
+	  plugin_cp_pre_genericize (fndecl);
 	  c_genericize (fndecl);
 	  c_gimple_diagnostics_recursively (fndecl);
 
diff -urp gcc-4.3.0-20080428-origplug/gcc/c-gimplify.c gcc-4.3.0-20080428/gcc/c-gimplify.c
--- gcc-4.3.0-20080428-origplug/gcc/c-gimplify.c	2008-07-10 14:19:27.000000000 +0200
+++ gcc-4.3.0-20080428/gcc/c-gimplify.c	2008-07-10 23:01:25.000000000 +0200
@@ -112,7 +112,10 @@ c_genericize (tree fndecl)
      the correct function.  */
   cgn = cgraph_node (fndecl);
   for (cgn = cgn->nested; cgn ; cgn = cgn->next_nested)
-    c_genericize (cgn->decl);
+    {
+      plugin_cp_pre_genericize (cgn->decl);
+      c_genericize (cgn->decl);
+    }
 }
 
 static void
